src/trieNode.ts

```diff
@@ -3,179 +3,195 @@ import { keccak256 } from 'ethereumjs-util'
   }
+}
 
-  // @ts-ignore
-  getKey(): number[] {
-    if (this.type !== 'branch') {
-      let key = this.raw[0]
-      let nibbles = removeHexPrefix(stringToNibbles(key))
-      return nibbles
-    }
+export class ExtensionNode {
+  _nibbles: Nibbles
+  _value: Buffer
+
+  constructor(nibbles: Nibbles, value: Buffer) {
+    this._nibbles = nibbles
+    this._value = value
+  }
+
+  static encodeKey(key: Nibbles): Nibbles {
+    return addHexPrefix(key, false)
+  }
+
+  static decodeKey(key: Nibbles): Nibbles {
+    return removeHexPrefix(key)
+  }
+
+  get key(): Nibbles {
+    return this._nibbles.slice(0)
+  }
+
+  set key(k: Nibbles) {
+    this._nibbles = k
+  }
+
+  get value(): Buffer {
+    return this._value
+  }
+
+  set value(v: Buffer) {
+    this._value = v
+  }
+
+  encodedKey(): Nibbles {
+    return ExtensionNode.encodeKey(this._nibbles.slice(0))
+  }
+
+  raw(): [Buffer, Buffer] {
+    return [nibblesToBuffer(this.encodedKey()), this._value]
   }
 
   serialize(): Buffer {
-    return rlp.encode(this.raw)
+    return rlp.encode(this.raw())
   }
 
   hash(): Buffer {
     return keccak256(this.serialize())
   }
+}
 
-  toString(): string {
-    // @ts-ignore
-    let out = NodeType[this.type]
-    out += ': ['
-    this.raw.forEach(function(el) {
-      if (Buffer.isBuffer(el)) {
-        out += el.toString('hex') + ', '
-      } else if (el) {
-        out += 'object, '
-      } else {
-        out += 'empty, '
-      }
-    })
-    out = out.slice(0, -2)
-    out += ']'
-    return out
-  }
-
-  getChildren(): (Buffer | number[])[][] {
-    var children = []
-    switch (this.type) {
-      case 'leaf':
-        // no children
-        break
-      case 'extention':
-        // one child
-        children.push([this.key, this.getValue()])
-        break
-      case 'branch':
-        for (let index = 0, end = 16; index < end; index++) {
-          const value = this.getValue(index)
-          if (value) {
-            children.push([[index], value])
-          }
-        }
-        break
-    }
-    return children
+export class LeafNode {
+  _nibbles: Nibbles
+  _value: Buffer
+
+  constructor(nibbles: Nibbles, value: Buffer) {
+    this._nibbles = nibbles
+    this._value = value
+  }
+
+  static encodeKey(key: Nibbles): Nibbles {
+    return addHexPrefix(key, true)
+  }
+
+  static decodeKey(encodedKey: Nibbles): Nibbles {
+    return removeHexPrefix(encodedKey)
+  }
+
+  get key(): Nibbles {
+    return this._nibbles.slice(0)
+  }
+
+  set key(k: Nibbles) {
+    this._nibbles = k
+  }
+
+  get value(): Buffer {
+    return this._value
+  }
+
+  set value(v: Buffer) {
+    this._value = v
+  }
+
+  encodedKey(): Nibbles {
+    return LeafNode.encodeKey(this._nibbles.slice(0))
+  }
+
+  raw(): [Buffer, Buffer] {
+    return [nibblesToBuffer(this.encodedKey()), this._value]
+  }
+
+  serialize(): Buffer {
+    return rlp.encode(this.raw())
+  }
+
+  hash(): Buffer {
+    return keccak256(this.serialize())
   }
 }
```

Comment:
Wouldn't it make sense to keep `TrieNode` as the parent class (or eventually an [abstract class](https://www.typescriptlang.org/docs/handbook/classes.html#abstract-classes)) and keep generic functionality like `serialize()` or `hash()` (and eventually some more) there?

Reply:
I tried something like that but the only method that can be moved there would be `hash()`. `serialize()` seems possible on first glance, but it calls `this.raw()` which has a different return type based on node type (specifically `BranchNode` is different than the other two).

Reply:
Ok.